src/dmi_decode.c

```diff
@@ -0,0 +1,250 @@
+        if (r == -1) {
+            if (errno != EINTR) {
+                perror(prefix);
+                return -1;
+            }
+        } else {
+            r2 += r;
+        }
+    }
+
+    if (r2 != count) {
+        fprintf(stderr, "%s: Unexpected end of file\n", prefix);
+        return -1;
+    }
+
+    return 0;
+}
+
+static void *read_file(off_t base, size_t *max_len, const char *filename)
+{
+    struct stat statbuf;
+    int fd;
+    uint8_t *p;
+
+    /*
+     * Don't print error message on missing file, as we will try to read
+     * files that may or may not be present.
+     */
+    if ((fd = open(filename, O_RDONLY)) == -1) {
+        if (errno != ENOENT)
+            perror(filename);
+
+        return NULL;
+    }
+
+    /*
+     * Check file size, don't allocate more than can be read.
+     */
+    if (fstat(fd, &statbuf) == 0) {
+        if (base >= statbuf.st_size) {
+            fprintf(stderr, "%s: Can't read data beyond EOF\n",
+                filename);
+            p = NULL;
+            goto out;
+        }
+        if (*max_len > (size_t)statbuf.st_size - base)
+            *max_len = statbuf.st_size - base;
+    }
+
+    if ((p = malloc(*max_len)) == NULL) {
+        perror("malloc");
+        goto out;
+    }
+
+    if (lseek(fd, base, SEEK_SET) == -1) {
+        fprintf(stderr, "%s: ", filename);
+        perror("lseek");
+        goto err_free;
+    }
+
+    if (att_read(fd, p, *max_len, filename) == 0)
+        goto out;
+
+err_free:
+    free(p);
+    p = NULL;
+
+out:
+    if (close(fd) == -1)
+        perror(filename);
+
+    return p;
+}
+
+static void ascii_filter(char *bp, size_t len)
+{
+    size_t i;
+
+    for (i = 0; i < len; i++)
+        if (!isprint(bp[i]))
+            bp[i] = '.';
+}
+
+static char *_dmi_string(const struct dmi_header *dm, uint8_t s, int filter)
+{
+    char *bp = (char *)dm->data;
+  
+    bp += dm->length;
+    while (s > 1 && *bp) {
+        bp += strlen(bp);
+        bp++;
+        s--;
+    }
+
+    if (!*bp)
+        return NULL;
+
+    if (filter)
+        ascii_filter(bp, strlen(bp));
+
+    return bp;
+}
+
+const char *dmi_string(const struct dmi_header *dm, uint8_t s)
+{
+    char *bp;
+  
+    if (s == 0)
+        return "Not Specified";
+  
+    bp = _dmi_string(dm, s, 1);
+    if (bp == NULL)
+          return bad_index;
+
+    return bp;
+}
+
+static int att_smbios3_decode(uint8_t *buf, const char *devmem, struct meta_data *meta)
+{
+    size_t len;
+    int ret = 0;
+    uint8_t *dmi_buf = NULL;
+
+    /* Don't let checksum run beyond the buffer */
+    if (buf[0x06] > 0x20) {
+        fprintf(stderr,
+            "Entry point length too large (%u bytes, expected %u).\n",
+            (unsigned int)buf[0x06], 0x18U);
+        return -1;
+    }
+
+    /* From specification: obtain offset for System Information */
+    len  = (uint32_t)((buf + 0x0C)[0] + ((buf + 0x0C)[1] << 8)
+                     + ((buf + 0x0C)[2] << 16) + ((buf + 0x0C)[3] << 24));
+ 
```

Comment:
In fact, since format of structure tables in`/sys/firmware/dmi/tables/DMI` is identical and we can get size of file in different ways, this check may not be necessary at all.

Reply:
I don't have any devices with SMBIOSv2 so I cannot check it.

Reply:
@0xDen @krystian-hebel I tested SMBIOSv2 on PC and it works.

I pushed commit enabling SMBIOSv2 and a small fix to CBOR encoding and attester sends metadata now.
I need to update Fobnail (it's metadata handling is a bit outdated) to check whether everything is ok.